util/pyclient/bft_client.py

```diff
@@ -241,3 +247,256 @@ def get_rsi_replies(self):
+        """async context manager method for 'with' statements"""
+        self.sock.close()
+
+    def __enter__(self):
+        """context manager method for 'with' statements"""
+        return self
+
+    def __exit__(self, *args):
+        """context manager method for 'with' statements"""
+        self.sock.close()
+        pass
+
+    def _process_received_msg_err(self, ex):
+        """In UDP, we do not except any errors while processing messages"""
+        # re-raise exception
+        raise ex
+
+class TcpTlsClient(BftClient):
+    """
+    Define a TCP/TLS client. This client communicates as a TLS over TCP client to all replicas.
+    It uses TCP to guarantee in-order data arrival and data reliability (using re-transmissions) and TLS to guarantee
+    authenticity and integrity. To enable a TLS-handshake, it uses self-signed SSL X.509 certificates.
+    """
+    # In create_tls_certs.sh - openssl command line utility uses CN(certificate name) in the subj field.
+    # This is the host name (domain name) to be verified.
+    CERT_DOMAIN_FORMAT="node%dser"
+    # Taken from TlsTCPCommunication.cpp (we prefer hard-code and not to parse the file)
+    MSG_HEADER_SIZE=4
+
+    def __init__(self, config, replicas):
+        super().__init__(config, replicas)
+        self.ssl_streams = dict()
+
+    async def on_started_replicas(self, replica_ids):
+        """
+        Establish a TLS over TCP connections to all replicas concurrently (using a nursery). 
+        We relay on caller cancel scope.
+        """
+        async with trio.open_nursery() as nursery:
+            for id in replica_ids:
+                replica = self.replicas[id]
+                if (replica.ip, replica.port) not in self.ssl_streams.keys():
+                    nursery.start_soon(self._establish_ssl_stream, replica)
+
+    def _get_private_key_path(self, replica_id, is_client=False):
+        """
+        Private key is under <certificate root path>/replica_id/<node type>/pk.pem,
+        where node type is "server" or "client".
+        """
+        cert_type = "client" if is_client else "server"
+        return os.path.join(self.config.certs_path, str(replica_id), cert_type, "pk.pem")
+
+    def _get_cert_path(self, replica_id, is_client=False):
+        """
+        Certificate is under <certificate root path>/replica_id/<node type>/cert.pem,
+        where node type is "server" or "client".
+        """
+        cert_type = "client" if is_client else "server"
+        return os.path.join(self.config.certs_path, str(replica_id), cert_type, cert_type + ".cert")
+
+    async def _establish_ssl_stream(self, dest_replica):
+        """
+        Try every 250ms to connect to dest_replica. We relay on the caller context that will trigger cancalleation at 
+        some point in time. Break the loop on success and keep the ssl_stream in self.ssl_streams for future use.
+        """
+        server_cert_path = self._get_cert_path(dest_replica.id, False)
```

Comment:
Leverage the optionality of the parameter: `server_cert_path = self._get_cert_path(dest_replica.id)`

Reply:
I made it explicit, same as in _get_cert_path. Actually, for now it is always False, but in case (for any reason) in the future we will have other IDs for clients, who might become TLS servers, ITs good to have an already make helper function. So I keep it with this option.